### DISCUSSION TITLE
How to Force M1 App to use rosetta

### DISCUSSION THREAD

**Discussion by 2snappy** [airsdk/Adobe-Runtime-Support#1791] [General]
My mac app has to use the HTMLLoader. Unfortunately, HTMLLoader doesn't work on the new apple M1 Chip computers. As such, I need to make sure that the rosetta checkbox is pre-selected for the mac app but I haven't been able to figure out how to compile the mac app so that it automatically has the rosetta checkbox selected.

Does anybody know how to build a Mac app that has the rosetta option pre-selected? Right now my clients have to manually select the rosetta option when they install the app otherwise things freeze when they open the htmlloader window.

**Comment by ajwfrost** [comment]
The only way I know of to force an app to run via Rosetta is to remove the support for arm64 from the main binary... which means that the main app and AIR runtime will be running in Rosetta mode (and via AS3 interpreter rather than JIT compiler).

You can use "lipo" for this - let me know if you need the details on this and we can have a quick try of it here, plus if anyone else needs the same sort of thing then we can perhaps add an option to the packaging process to choose between universal binary and intel-only binary..

FYI we've also removed the WebKit libraries and adapted HTMLLoader from the newest AIR SDK releases, which means that they will work fine on M1 macs, although the HTML capabilities are somewhat changed....

thanks

**Reply by 2snappy** [reply]
Thanks for your comments. They were very helpful.

To anyone else that wants to remove the arm64 from your app so that it defaults to using the Rosetta mode here are the lipo commands.

To verify that your app has both architectures you can do the following command in the folder where your app is located. You would replace $APP with the name of your app.

**lipo -info "$APP.app/Contents/MacOS/$APP"**

To remove the arm64 architecture from your app you would use the following command.

**lipo -remove arm64 "$APP.app/Contents/MacOS/$APP" -output "$APP.app/Contents/MacOS/$APP"**

To verify that it is removed you would just use this command again

**lipo -info "$APP.app/Contents/MacOS/$APP"**

**Comment by porteusconf** [comment]
Why not just tell users how to force the app to launch in rosetta2?  That is, I think lipo is not the only way to force rosetta2.
I think there are much better ways to force rosetta2, including: Select App in finder-> File-> Get Info-> check box for "Run with Rosseta). 

In the general case, I am pretty sure an even better way is if you launch a rosetta2 terminal (or iterm), then every child-process of that rosseta2 terminal will attempt to run with same x86_64 arch rather than arm64. This was the method used to install homebrew on M1 macs before homebrew supported M1 natively.

I think this method works since preference for x86_64 rather than arm64 is inherited by every child process. So this method, detailed below, you need not do lipo (suction) which seems to me a bad idea anyways (the app you did lipo on will just become fat again on the next upgrade/update). And no need to run around putting checks in the "Use Rosseta box" either, except in just one customized Terminal app (or iTerm).

See https://news.ycombinator.com/item?id=25132217

child processes will prefer the arch of the parent process if one is available to maximize compatibility. So for example launching x86_64 sh will cause a python invocation from that $
The "arch" utility is your escape hatch to switch the preferred arch for the spawned process and all of its children.

In a nutshell,  create a copy of Terminal (or iTerm), and rename it TermRosetta2 (or similar). For that app, check the box that says "Open with Rosseta2".  In Terminal use any of the commands below to launch apps in reosetta2 mode. Tab-completion only works on non-Apple applications btw...
```
open  /Applications/MyApp.app
open /Applications/Google\ Chrome.app
/Applications/MyApp.app/Contents/Macos/MyApp
/Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome 
arch -x86_64 /Applications/Google\ Chrome.app/Contents/MacOS/Google\ Chrome 
```
Any/all of the above commands should launch apps in rosetta mode.
